File: benchmarks/11-special-functions/special_functions.m
(* ==========================================================================
   Experiment 11 -- Special functions over machine arrays
   ==========================================================================
   WHAT IT MEASURES.  src/special_functions/ (41 files) reached through the
   packed-array path -- src/ndkernels.c and sf_machine.c.  Unlike group A this
   is an EXECUTION comparison: scipy.special is compiled C, so a gap here is
   overhead or a missing vector kernel, not a missing algorithm.

   Sizes are 10^6, comfortably above PACK_MIN_ELEMENTS (src/pack.h:84), so the
   buffer path is the one being measured.

   Checks are a rounded scalar from a small DETERMINISTIC input -- never a sum
   over the random timing data, which the three systems cannot align.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"Gamma", "LogGamma", "BesselJ", "BesselY", "Erf", "Erfc",
         "PolyGamma", "AiryAi", "Zeta", "ProductLog", "EllipticK", "Beta"}];

n = 1000000;
v = rand01[{n}] + 1/2;              (* in [0.5, 1.5]: safe for every function *)

bench["Gamma over 10^6", Gamma[v];];
check["Gamma over 10^6", Round[10^6 N[Gamma[3/2]]]];

bench["Erf over 10^6", Erf[v];];
check["Erf over 10^6", Round[10^6 N[Erf[1/2]]]];

bench["BesselJ[0, .] over 10^6", BesselJ[0, v];];
check["BesselJ[0, .] over 10^6", Round[10^6 N[BesselJ[0, 1/2]]]];

bench["LogGamma over 10^6", LogGamma[v];];
check["LogGamma over 10^6", Round[10^6 N[LogGamma[3/2]]]];

bench["PolyGamma[0, .] over 10^6", PolyGamma[0, v];];
check["PolyGamma[0, .] over 10^6", Round[10^6 N[PolyGamma[0, 3/2]]]];

benchIf["AiryAi over 10^6", "AiryAi", AiryAi[v];];
checkIf["AiryAi over 10^6", "AiryAi", Round[10^6 N[AiryAi[1/2]]]];

benchIf["Zeta over 10^6", "Zeta", Zeta[v + 1];];
checkIf["Zeta over 10^6", "Zeta", Round[10^6 N[Zeta[5/2]]]];

File: benchmarks/11-special-functions/special_functions.py
#!/usr/bin/env python3
"""Experiment 11 -- Special functions over machine arrays (scipy column).

Same kernels as ``special_functions.m``, same order and sizes.

Unlike group A this is an EXECUTION comparison: scipy.special is compiled C, so
a gap here is overhead or a missing vector kernel, not a missing algorithm.
Sizes are 10**6, above PACK_MIN_ELEMENTS, so Mathilda's buffer path is what is
being measured.

Checks are a rounded scalar from a small deterministic input -- never a sum over
the random timing data, which the three systems cannot align.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
import scipy.special as sp

from harness import bench, check, require, seed

require(["scipy.special:gamma", "scipy.special:erf", "scipy.special:jv",
         "scipy.special:gammaln", "scipy.special:psi", "scipy.special:airy",
         "scipy.special:zeta"])

seed()
n = 1000000
v = np.random.random(n) + 0.5          # [0.5, 1.5]: safe for every function


def r6(x):
    return int(np.floor(x * 1e6 + 0.5))


bench("Gamma over 10^6", lambda: sp.gamma(v))
check("Gamma over 10^6", r6(sp.gamma(1.5)))

bench("Erf over 10^6", lambda: sp.erf(v))
check("Erf over 10^6", r6(sp.erf(0.5)))

bench("BesselJ[0, .] over 10^6", lambda: sp.jv(0, v))
check("BesselJ[0, .] over 10^6", r6(sp.jv(0, 0.5)))

bench("LogGamma over 10^6", lambda: sp.gammaln(v))
check("LogGamma over 10^6", r6(sp.gammaln(1.5)))

bench("PolyGamma[0, .] over 10^6", lambda: sp.psi(v))
check("PolyGamma[0, .] over 10^6", r6(sp.psi(1.5)))

bench("AiryAi over 10^6", lambda: sp.airy(v)[0])
check("AiryAi over 10^6", r6(sp.airy(0.5)[0]))

bench("Zeta over 10^6", lambda: sp.zeta(v + 1))
check("Zeta over 10^6", r6(sp.zeta(2.5)))
